// redis-smq/dist/src/lib/message-manager/messages/pending-messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PendingMessages = void 0;
const types_1 = require("../../../../types");
const queue_1 = require("../../queue-manager/queue");
const pending_priority_messages_1 = require("./pending-priority-messages");
const pending_non_priority_messages_1 = require("./pending-non-priority-messages");
class PendingMessages {
    constructor(config, redisClient, logger) {
        this.redisClient = redisClient;
        this.config = config;
        this.pendingNonPriorityMessages = new pending_non_priority_messages_1.PendingNonPriorityMessages(config, redisClient, logger);
        this.pendingPriorityMessages = new pending_priority_messages_1.PendingPriorityMessages(config, redisClient, logger);
    }
    purge(queue, cb) {
        const queueParams = queue_1.Queue.getParams(this.config, queue);
        queue_1.Queue.getSettings(this.config, this.redisClient, queueParams, (err, settings) => {
            if (err)
                cb(err);
            else if ((settings === null || settings === void 0 ? void 0 : settings.type) === types_1.EQueueType.PRIORITY_QUEUE) {
                this.pendingPriorityMessages.purge(queueParams, cb);
            }
            else {
                this.pendingNonPriorityMessages.purge(queueParams, cb);
            }
        });
    }
    list(queue, skip, take, cb) {
        queue_1.Queue.getSettings(this.config, this.redisClient, queue, (err, settings) => {
            if (err)
                cb(err);
            else if ((settings === null || settings === void 0 ? void 0 : settings.type) === types_1.EQueueType.PRIORITY_QUEUE) {
                this.pendingPriorityMessages.list(queue, skip, take, cb);
            }
            else {
                this.pendingNonPriorityMessages.list(queue, skip, take, cb);
            }
        });
    }
    delete(queue, messageId, sequenceId, cb) {
        queue_1.Queue.getSettings(this.config, this.redisClient, queue, (err, settings) => {
            if (err)
                cb(err);
            else if ((settings === null || settings === void 0 ? void 0 : settings.type) === types_1.EQueueType.PRIORITY_QUEUE) {
                this.pendingPriorityMessages.delete(queue, messageId, cb);
            }
            else {
                this.pendingNonPriorityMessages.delete(queue, messageId, sequenceId, cb);
            }
        });
    }
    count(queue, cb) {
        queue_1.Queue.getSettings(this.config, this.redisClient, queue, (err, settings) => {
            if (err)
                cb(err);
            else if ((settings === null || settings === void 0 ? void 0 : settings.type) === types_1.EQueueType.PRIORITY_QUEUE)
                this.pendingPriorityMessages.count(queue, cb);
            else
                this.pendingNonPriorityMessages.count(queue, cb);
        });
    }
}
exports.PendingMessages = PendingMessages;
