// redis-smq/dist/types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EExchangeType = exports.EQueueType = exports.EQueueSettingType = exports.EMessageUnacknowledgedCause = exports.EMessageDeadLetterCause = void 0;
var EMessageDeadLetterCause;
(function (EMessageDeadLetterCause) {
    EMessageDeadLetterCause["TTL_EXPIRED"] = "ttl_expired";
    EMessageDeadLetterCause["RETRY_THRESHOLD_EXCEEDED"] = "retry_threshold_exceeded";
    EMessageDeadLetterCause["PERIODIC_MESSAGE"] = "periodic_message";
})(EMessageDeadLetterCause = exports.EMessageDeadLetterCause || (exports.EMessageDeadLetterCause = {}));
var EMessageUnacknowledgedCause;
(function (EMessageUnacknowledgedCause) {
    EMessageUnacknowledgedCause["TIMEOUT"] = "timeout";
    EMessageUnacknowledgedCause["CONSUME_ERROR"] = "consumer_error";
    EMessageUnacknowledgedCause["UNACKNOWLEDGED"] = "unacknowledged";
    EMessageUnacknowledgedCause["OFFLINE_CONSUMER"] = "offline_consumer";
    EMessageUnacknowledgedCause["TTL_EXPIRED"] = "ttl_expired";
})(EMessageUnacknowledgedCause = exports.EMessageUnacknowledgedCause || (exports.EMessageUnacknowledgedCause = {}));
var EQueueSettingType;
(function (EQueueSettingType) {
    EQueueSettingType["PRIORITY_QUEUING"] = "PRIORITY_QUEUING";
    EQueueSettingType["QUEUE_TYPE"] = "QUEUE_TYPE";
    EQueueSettingType["RATE_LIMIT"] = "RATE_LIMIT";
    EQueueSettingType["EXCHANGE"] = "EXCHANGE";
})(EQueueSettingType = exports.EQueueSettingType || (exports.EQueueSettingType = {}));
var EQueueType;
(function (EQueueType) {
    EQueueType[EQueueType["LIFO_QUEUE"] = 0] = "LIFO_QUEUE";
    EQueueType[EQueueType["FIFO_QUEUE"] = 1] = "FIFO_QUEUE";
    EQueueType[EQueueType["PRIORITY_QUEUE"] = 2] = "PRIORITY_QUEUE";
})(EQueueType = exports.EQueueType || (exports.EQueueType = {}));
var EExchangeType;
(function (EExchangeType) {
    EExchangeType[EExchangeType["DIRECT"] = 0] = "DIRECT";
    EExchangeType[EExchangeType["FANOUT"] = 1] = "FANOUT";
    EExchangeType[EExchangeType["TOPIC"] = 2] = "TOPIC";
})(EExchangeType = exports.EExchangeType || (exports.EExchangeType = {}));
